shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,790 @@
-  var stackedUsers = {};
-  stackedUsers[openerIdentityId] = true;
-  var userStack = [openerIdentityId];
-
-  while (userStack.length > 0) {
-    var recipient = userStack.pop();
-    var edges = edgesByRecipient[recipient];
-    if (edges) {
-      for (var ii = 0; ii < edges.length; ++ii) {
-        var inEdge = edges[ii];
-        var sharer = inEdge.sharer;
-        if (sharer == owner) {
-          return true;
+  const stack = [chooseAll];
+  // Array of Map<hashedClause, { clause: Clause, tokensUsed: [String] }>.
+
+  while (stack.length > 0) {
+    const chosen = stack.pop();
+    let minimal = true;
+    for (const removedHashedClause in chosen) {
+      const newChosen = {};
+      for (const hashedClause in chosen) {
+        if (hashedClause !== removedHashedClause) {
+          newChosen[hashedClause] = chosen[hashedClause];
         }
-        if (!stackedUsers[sharer]) {
-          userStack.push(sharer);
-          stackedUsers[sharer] = true;
+      }
+
+      if (fullCover(newChosen)) {
+        minimal = false;
+        // TODO(perf): keep track of values we've already explored.
+        stack.push(newChosen);
+      }
+    }
+
+    if (minimal) {
+      const resultClause = new Clause();
+      const tokensUsed = {};
+      for (const hashedClause in chosen) {
+        resultClause.conjoin(chosen[hashedClause].clause);
+        for (const tokenId in chosen[hashedClause].tokensUsed) {
+          tokensUsed[tokenId] = true;
         }
       }
+
+      result[resultClause.hash()] = { clause: resultClause, tokensUsed: tokensUsed };
+    }
+  }
+
+  return result;
+}
+
+function proveClauses(db, context, goalClauses) {
+  // `goalClauses` is of the form of a result of `normalize()`, that is:
+  // `Map<ClauseId, {clause: Clause, tokensUsed: [tokenId]}>`.
+  //
+  // This function attempts to determine whether at least one of the input clauses holds.
+  // If a proof is found, returns `{yes: {tokensUsed: [tokenId]}}`.
+  // otherwise, returns {no: {}}
+  //
+  // The `db` parameter is optional. If it is present, the database will be queried as needed.
+  // If left out, the computation will proceed with only the tokens already present in `context`.
+
+  const clauses = {};
+  const clauseStack = []; // clause IDs
+
+  for (const hashedClause in goalClauses) {
+    const goalClause = goalClauses[hashedClause];
+    const tokensUsed = goalClause.tokensUsed;
+    if (goalClause.clause.isEmpty()) {
+      // We have an empty goal. Trivially true.
+      return { yes: { tokensUsed: tokensUsed } };
+    }
+
+    clauses[hashedClause] = { clause: goalClause.clause, tokensUsed: tokensUsed };
+    clauseStack.push(hashedClause);
+  }
+
+  const grainsFullyQueried = {};
+
+  while (clauseStack.length > 0) {
+    const hashedClause = clauseStack.pop();
+    const clause = new Clause();
+    clause.conjoin(clauses[hashedClause].clause);
+    const tokensUsed = clauses[hashedClause].tokensUsed;
+    const goal = clause.popFirstGoal();
```

Comment:
Also, I think goals could be cyclic. Of course, a goal which is implied by itself should be treated as false (think of it as: all goals are false until shown to be true).

I think this situation gets complicated, though:

A <= B or C
B <= A
C <= true

A naive algorithm trying to prove A might start by following the branch to B, which leads back to A, which loops.

A better but still naive algorithm might follow the branch to B, see that it leads back to A, and so decide that B is false because cycles are false, and memoize this. It may then try to prove C, find that it's true, and therefore decide that A is true. However, B has now been incorrectly memoized as false, when in the end it should have been true.

A better algorithm might have three results: "true", "false", and "cycle", where "cycle" is a false result but is never memoized. This might work? Will not memoizing mean repeating work too often?

I'm sure this problem has been deeply studies, of course, so maybe we should look up the solution.